sisl/physics/tests/test_hamiltonian.py

```diff
@@ -589,16 +589,23 @@ def test_eig4(self, setup):
             assert np.allclose(e, es.eig)
             assert np.allclose(v, es.state.T)
             assert np.allclose(es.norm2(), 1)
-            assert np.allclose(es.inner(diagonal=False) - np.eye(len(es)), 0)
+            assert np.allclose(es.inner(diag=False), np.eye(len(es)))
 
-            assert es.inner(es.sub(0)).shape == (1, )
-            assert es.inner(es.sub(0), diagonal=False).shape == (len(es), 1)
+            assert es.inner(es.sub(0)).shape == (len(es), )
```

Comment:
Note that here the number of bras and kets differ - what is then the meaning of asking just for the diagonal part (the default)?

Reply:
Agreed... This is not clear. However, currently it just returns the first `n` diagonal components where `n` is the lowest dimension of the `left/right` state vectors. Agreed it does not make sense... Perhaps we should issue a warning in these cases, but then again, users would probably force `diag=False` in those cases since that would be what they want... Hmmm...

Reply:
Sounds good to issue a warning when `diag=True` and `right` is provided.

Reply:
something like this:
```python
 warnings.warn(f"{self.__class__.__name__}.inner matrix product is non-square, only the first {min(len(self), len(right))} diagonal elements will be returned")
```